src/datasets/packaged_modules/imagefolder/imagefolder.py

```diff
@@ -124,19 +120,36 @@ def analyze(files_or_archives, downloaded_files_or_dirs, split):
             if do_analyze:
                 analyze(files, downloaded_files, split_name)
                 analyze(archives, downloaded_dirs, split_name)
+
+                if metadata_files:
+                    add_metadata = not (self.config.drop_metadata is True)
+                    add_labels = self.config.drop_labels is False
+                else:
+                    add_metadata = False
+                    add_labels = not (self.config.drop_labels is True)
```

Comment:
I just added this logic, I hope it helps making things clearer.

I also updated the tests you mentioned @polinaeterna : 
- test_generate_examples_with_metadata_that_misses_one_image
- test_generate_examples_with_metadata_in_wrong_location
- test_generate_examples_drop_metadata

Feel free to revert if you think it's not the right solution.

Reply:
I like it that way, thank you! 

So the default behavior (both `drop_metadata` and `drop_labels` are None) is: 
parse metadata and labels, if `metadatata_files` are found - add metadata and do not add inferred labels even if they are found. Add labels only if it’s explicitly specified by a user (`drop_labels` is set to False)

If metadata files are not found, default behavior is to add labels if they are found.

And if we add both inferred labels and metadata, if "label" column is found in metadata, it replaces the inferred labels. 

Right?

Reply:
ah no, I see that it raises an error if "label" feature is present both in `metadata.jsonl` and added if `add_labels` is True here: https://github.com/huggingface/datasets/pull/4622/files#diff-8bb31057be49645ccb98250431e4ca7e663d1b3413c017a4c932f0c8b77b8188R194

do you think we should preserve this behavior? if so, ignore my comments below

Reply:
Yea the rule is:
- if there are metadata, use the metadata
- otherwise, infer labels from the file names
- if the user is not interested by the provided metadata or the labels, they can explicitly drop them

Then there how to get labels from a dataset with metadata, and in this case the user can again explicitly ask for the labels with drop_labels=False (but this is a rare case IMO). And in this case, I think the "label" column should override the "label" column in the metadata (otherwise the user can simply not ask for the labels).

Therefore I would keep the current behavior:
```python
{..., **image_metadata, **image_label}
```

Reply:
ok, agree! but current behavior in cases when "label" is both inferred from data files (and `drop_labels`=False) and is present in metadata features (and `drop_metadata` is None or False) is to raise an error:
```
ValueError: Metadata feature keys ['label'] are already present as the image features
```

Another option would be to give a warning in this case and add to `self.config.features` only those features that are not already present in `self.config.features` (because we want "label" column to be for sure a ClassLabel but it might be inferred from metadata as a string). But I agree it seems to be a corner case (though I'm not an expert in image processing at all)

Reply:
@mariosasko what do you think about the case described above? do you think we should preserve the current behavior and raise error when “label” is presented both in metadata and inferred from filenames (while drop_labels=False)? I personally would prefer to log a warning and chose labels from filenames but it’s not that important at all and we can leave as is if you agree on that.
I added some comments to the code and modified ImageFolder documentation, if you think they’re ok I guess we can merge?

Reply:
`drop_labels=False` means the labels from filenames are explicitly requested, so a warning sounds good. Please also add a test for that.